workspace/src/localization/localization_py/localization_py/particle_filter.py

```diff
@@ -0,0 +1,180 @@
+
+        Returns:
+            A 4-D state for the vehicle, consisting of x, y, theta, and velocity v.
+
+        """
+
+        self.timestep = self.timestep+1
+        #first, update the motion model
+        for i in range(0, self.num_particles):
+            myu = np.array([[u[0,0]+np.random.normal(0,0.05)], [u[1,0]+np.random.normal(0,0.05)]])
+            self.particles[i] = self.dyn.motion_model(self.particles[i], myu)
+        self.update_dist(obs)
+        #then, reassign weights in relation to the observation
+        self.assign_weights()
+
+        state = np.zeros((4,1))
+        for i in range(0,self.num_particles):
+            for j in range(0,4):
+                state[j,0] = state[j,0]+self.particles[i][j,0]*self.particle_weights[i]
+        if(self.timestep%5 == 0):
+            self.resample()
+        #then, return final value
+        return state
+    
+    def update_dist(self,obs):
+        """Updates each particle error distribution from the observation.
+
+        For each particle, compute the distance between the current state and the observation (position and heading). Add this to the associated distribution, and pop the oldest value.
+        
+        Args:
+            obs: The 3-D observation vector.
+        """
+
+        for i in range(0, self.num_particles):
+            distance = math.sqrt((self.particles[i][0,0]-obs[0,0])**2+(self.particles[i][1,0]-obs[1,0])**2)
+            self.particle_distr_dist[i].append(distance)
+            self.particle_distr_head[i].append(self.particles[i][2,0]-obs[2,0])
+            if(len(self.particle_distr_dist[i])>10):
+                self.particle_distr_dist[i].pop(0)
+                self.particle_distr_head[i].pop(0)
+
+    def resample(self):
+        """Resample the particles for the next generation.
+
+        Generates a new generation of particles. Computes a cummulative weight, and then generates num_particles random numbers. Then, whichever cummulative range the random number lies in is the particle that gets propogated. If a particle gets propogated more than once, some randomness is added to the state. 
+
+        """
+
+        cumulative_weights = [sum(self.particle_weights[:i+1]) for i in range(self.num_particles)]
+        sampled_points = []
+        for _ in range(self.num_particles):
+            random_num = random.random()  # Generate a random number between 0 and 1
+
+            # Find the interval where the random number falls
+            for i, cum_weight in enumerate(cumulative_weights):
+                if random_num < cum_weight:
+                    sampled_points.append(i)
+                    break
+        already_seen = []
+        new_particles = []
+        new_pdd = []
+        new_pdh = []
+        new_hpx= []
+        new_hpy = []
+        for j in range(0,self.num_particles):
```

Comment:
If you use numpy arrays for all these quantities, this entire loop gets eliminated. 

We seem to be using numpy arrays for the wrong things Ex. for x which is a 4x1 array in which you anyways update values individually.

Reply:
Here we use a list because this is somewhat dynamic... There are some versions of this where the number of particles throughout a single run is dynamic (so it is nice to be able to append/pop elements from these lists). In a c/c++ version of this it would definitely be cleaner, but for now it is fast enough...